spring-cloud-dataflow-admin/src/main/java/org/springframework/cloud/dataflow/admin/controller/StreamController.java

```diff
@@ -222,34 +221,85 @@ private void deployStream(StreamDefinition stream, Map<String, String> cumulated
+		StreamDefinition.DeploymentOrderIterator iterator = stream.getDeploymentOrderIterator();
+		while (iterator.hasNext()) {
 			ModuleDefinition currentModule = iterator.next();
-			boolean isFirst = !iterator.hasNext();
-			boolean isLast = (i == 0);
-			ModuleType type = determineModuleType(currentModule, isFirst, isLast);
-			ModuleRegistration registration = this.registry.find(currentModule.getName(), type);
-			if (registration == null) {
-				throw new IllegalArgumentException(String.format(
-						"Module %s of type %s not found in registry", currentModule.getName(), type));
-			}
-			ModuleCoordinates coordinates = registration.getCoordinates();
+
+			ModuleType type = determineModuleType(currentModule, !iterator.hasMoreUpstream(), !iterator.hasMoreDownstream());
+			ModuleCoordinates coordinates = getModuleCoordinates(type, currentModule);
+
 			Map<String, String> moduleDeploymentProperties = getModuleDeploymentProperties(currentModule, cumulatedDeploymentProperties);
-			boolean upstreamModuleSupportsPartition = upstreamModuleHasPartitionInfo(stream, currentModule, cumulatedDeploymentProperties);
-			// consumer module partition properties
-			if (isPartitionedConsumer(currentModule, moduleDeploymentProperties, upstreamModuleSupportsPartition)) {
-				updateConsumerPartitionProperties(moduleDeploymentProperties);
-			}
-			// producer module partition properties
-			if (isDownStreamModulePartitioned) {
-				updateProducerPartitionProperties(moduleDeploymentProperties, nextModuleCount);
-			}
+
+			downstreamInfo = postProcessPartitioningProperties(cumulatedDeploymentProperties, downstreamInfo, currentModule, iterator.peekUpstream(), moduleDeploymentProperties);
+
 			this.deployer.deploy(new ModuleDeploymentRequest(currentModule, coordinates, moduleDeploymentProperties));
-			nextModuleCount = getNextModuleCount(moduleDeploymentProperties);
-			isDownStreamModulePartitioned = isPartitionedConsumer(currentModule, moduleDeploymentProperties,
-					upstreamModuleSupportsPartition);
+		}
+	}
+
+	/**
+	 * Maybe update the {@literal currentModule} {@literal moduleDeploymentProperties} with properties related
+	 * to partitioning.
+	 *
+	 * @param cumulatedDeploymentProperties the whole set of stream deployment properties
+	 * @param downstreamInfo information about the downstream module in the stream
+	 * @param currentModule the module whose deployment properties we want to alter
+	 * @param upstreamModule the next module after {@literal currentModule} in deployment order, or null if none
+	 * @param moduleDeploymentProperties the deployment properties for the current module
+	 * @return information about the current module, to be used as 'previous' in the next invocation
+	 */
+	private ModulePartitionInfo postProcessPartitioningProperties(Map<String, String> cumulatedDeploymentProperties, ModulePartitionInfo downstreamInfo, ModuleDefinition currentModule, ModuleDefinition upstreamModule,  Map<String, String> moduleDeploymentProperties) {
```

Comment:
this method does too much - not only (potentially) updating the properties, but also returning a `ModulePartitionInfo` instance